From a3b078c4bd6f9d612379122e18e6e35d67c87222 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Thu, 22 Nov 2018 17:03:12 +1300
Subject: [PATCH] TRITON-779 store affinity rules when provisioning Reviewed
 by: Todd Whiteman <todd.whiteman@joyent.com> Approved by: Todd Whiteman
 <todd.whiteman@joyent.com>

---
 lib/endpoints/vms.js  |  7 +++++++
 package.json          |  2 +-
 test/vms.full.test.js | 21 +++++++++++++++++++--
 3 files changed, 27 insertions(+), 3 deletions(-)

diff --git a/lib/endpoints/vms.js b/lib/endpoints/vms.js
index 3d37dff..529b274 100644
--- a/lib/endpoints/vms.js
+++ b/lib/endpoints/vms.js
@@ -1300,6 +1300,13 @@ function createVm(req, res, next) {
                 req.params.sync = (sync === 'true' ? true : false);
             }
 
+            var locality = req.params.locality;
+            if (locality) {
+                var metadata = req.params.internal_metadata || {};
+                metadata.locality = JSON.stringify(locality);
+                req.params.internal_metadata = metadata;
+            }
+
             done();
         });
     }
diff --git a/package.json b/package.json
index de3dbb1..95e3989 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "vmapi",
   "description": "VMs API",
-  "version": "9.6.4",
+  "version": "9.6.5",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
diff --git a/test/vms.full.test.js b/test/vms.full.test.js
index 12a0718..c3112de 100644
--- a/test/vms.full.test.js
+++ b/test/vms.full.test.js
@@ -900,7 +900,11 @@ exports.create_vm = function (t) {
         ram: 64,
         quota: 10,
         customer_metadata: md,
-        creator_uuid: CUSTOMER
+        creator_uuid: CUSTOMER,
+        locality: {
+            strict: false,
+            far: [ '00000000-0000-0000-0000-000000000001' ]
+        }
     };
 
     var opts = createOpts('/vms', vm);
@@ -950,7 +954,6 @@ exports.wait_provisioned_job = function (t) {
 };
 
 
-
 exports.check_create_vm_nics_running = function (t) {
     var query = {
         belongs_to_uuid: newUuid,
@@ -964,6 +967,20 @@ exports.check_create_vm_nics_running = function (t) {
 };
 
 
+exports.check_locality_in_internal_metadata = function (t) {
+    client.get(vmLocation, function getVm(err, req, res, body) {
+        common.ifError(t, err);
+
+        t.deepEqual(JSON.parse(body.internal_metadata.locality), {
+            strict: false,
+            far: [ '00000000-0000-0000-0000-000000000001' ]
+        });
+
+        t.done();
+    });
+};
+
+
 exports.stop_vm = function (t) {
     var params = {
         action: 'stop'
-- 
2.21.0

